[Marcus Isaac Goh Wei Jie] iP

Marcusgwj opened the pull request:
### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/Marcusgwj/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [x] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
``` java
public class Main {
    public static void main(String[] args) { 
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Marcusgwj committed:
Level 1. Greet, Echo, Exit

Marcusgwj committed:
Level 2. Add, List

Marcusgwj committed:
Level 3. Mark as Done

Marcusgwj committed:
Level 4. ToDos, Events, Deadlines

Marcusgwj committed:
Automated Text UI Testing

Marcusgwj committed:
Level 5. Handle Errors

Marcusgwj committed:
Level 6. Delete

Marcusgwj committed:
A-Enums

Marcusgwj committed:
Fixed EXPECTED.TXT file error for A-TextUiTesting

Marcusgwj committed:
Add saving and loading of tasks to hard disk (Level-7)

Marcusgwj committed:
Add Dates and Times (Level 8)

Marcusgwj committed:
Merge branch 'branch-Level-7'

Marcusgwj committed:
Merge branch 'branch-Level-8'

Marcusgwj committed:
Add more OOP

Marcusgwj committed:
Organize into Packages

Marcusgwj committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Marcusgwj committed:
Modify Gradle configuration

Marcusgwj committed:
Add JUnit tests

Marcusgwj committed:
Add Tasklist test cases

Marcusgwj committed:
Add JavaDoc comments

Marcusgwj committed:
Merge branch 'branch-A-JavaDoc'

Marcusgwj committed:
Tweak code to comply with coding standard

Marcusgwj committed:
Merge branch 'branch-A-CodingStandard'

Marcusgwj committed:
Add ability to find tasks by searching for a keyword

Marcusgwj committed:
Merge branch 'branch-Level-9'

Marcusgwj committed:
Update JavaDoc for FindCommand

Marcusgwj commented:
### DukePro
> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/Marcusgwj/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) { 
        Application.launch(MainApp.class, args);
    }
}
```

Marcusgwj committed:
Use CheckStyle

Marcusgwj committed:
Merge branch 'branch-A-CheckStyle'

Marcusgwj committed:
Add a GUI to Duke

Marcusgwj committed:
Merge branch 'branch-Level-10'

Marcusgwj committed:
Use Varargs

Marcusgwj committed:
Merge branch 'branch-A-Varargs'

neosunhan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public static Command parseCommand(String userInput, TaskList taskList) {
+        try {
+            String [] inputArr = userInput.split(" ", 2);
```
```suggestion
            String[] inputArr = userInput.split(" ", 2);
```

neosunhan commented on src/main/java/duke/Response.java:
```diff
@@ -0,0 +1,43 @@
+package duke;
+
+/**
+ * A duke.Response from the application to some user input.
```
I think there's no need to include the package of the `Response` class.
```suggestion
 * A Response from the application to some user input.
```
Same goes for the other JavaDoc comments in this file.

neosunhan commented on src/main/java/duke/Response.java:
```diff
@@ -0,0 +1,43 @@
+     */
+    @Override
+    public String toString() {
+        return this.message;
```
> Avoid unnecessary use of `this` with fields.
> Use the `this` keyword only when a field is shadowed by a method or constructor parameter.

This is an advanced rule of the coding standard which you may want to follow. I noticed this in a few other places too.

neosunhan commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,17 @@
+package duke.command;
+
+import duke.Response;
+import duke.exception.DukeException;
+/**
```
```suggestion
import duke.exception.DukeException;

/**
```

neosunhan commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,17 @@
+     * @return The Response to be displayed.
+     * @throws DukeException If the action cannot be executed.
+     */
+    public abstract Response action() throws DukeException;
```
Since method names should be verbs, perhaps you can consider renaming this method to something like `execute`.

neosunhan commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,49 @@
+        String date = descriptionDate[1];
+        Event event = new Event(description, date);
+        this.taskList.addTask(event);
+        return new Response("Got it. I've added this task: " + "\n"
+                + event + "\n"
+                + "Now you have " + this.taskList.getSize() + " tasks in the list." + "\n");
```
I think some of these strings can be combined. If I recall correctly, string concatenation in Java is O(length of string) so unnecessary use of the + operator should be avoided. For example the `"\n"` can just be added to the end of the previous string. I've noticed this in several of the other files too.

neosunhan reviewed (commented):
LGTM. This code follows the Java coding standard very closely. You may want to consider following some of the advanced rules. Overall, I think the code is very readable. Variable and method names are also well chosen.

Marcusgwj committed:
Update code to follow coding standard

connlim commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,36 @@
+     * @param message The message to be printed.
+     */
+    public void printMessage(String message) {
+        System.out.println(BORDER + message + BORDER);
+    }
+
+    /**
+     * Prints a welcome message to the user.
+     */
+    public void welcomeMessage() {
+        System.out.println(BORDER + GREETING + BORDER);
+    }
+
+    /**
+     * Prints an error message to the user.
+     *
+     * @param error The error message to be printed.
+     */
+    public void invalidMessage(String error) {
+        System.out.println(BORDER + INVALID + error + "\n" + BORDER);
```
Duplication of border printing code. Consider extracting into a private method that takes in a String and prints the String along with the borders.

connlim commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,117 @@
+        this.description = description;
+        this.isDone = false;
+    }
+    /**
```
Add a space between methods

connlim commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,117 @@
+        Task task;
+
+        switch (taskType) {
+        case DEADLINE:
```
Duplicate code as it is possible to wrap the whole switch case in the same catch block.

connlim reviewed (commented):
Good job!

Marcusgwj committed:
Add assertions in Command classes

DeleteCommand, FindCommand, MarkCommand UnmarkCommand classes
updates a Task/queries for a List of Tasks.

There may be a possibility that it fails to update/query and
return null.

Assertions are added to ensure that updates/queries are done
successfully and non-null Task/List of Tasks is returned.

Assertions added to following classes:
* DeleteCommand
* FindCommand
* MarkCommand
* UnmarkCommand

Marcusgwj committed:
Merge pull request #1 from Marcusgwj/branch-A-Assertions

Add assertions in Command classes

Marcusgwj committed:
Improve Code Quality

Unnecessary this keywords is used.

Removing unnecessary this keywords.

Improving code quality maximises readability

Marcusgwj committed:
Merge pull request #2 from Marcusgwj/branch-A-CodeQuality

Improve Code Quality

Marcusgwj committed:
Add ability to tag tasks

All tasks are grouped together.

Adding tags allow users can classify tasks using tags.

Let's,
* allow users to tag tasks according to different categories.
* add ability to add and remove multiple tags for a task.
* update find feature to allow users to find by tags.

Marcusgwj committed:
Merge pull request #3 from Marcusgwj/branch-C-Tagging

Add ability to tag tasks

Marcusgwj committed:
Create gradle.yml

Set up CI

Marcusgwj committed:
Fix delimiter error for Tasks

Marcusgwj committed:
Add user guide
